tests/components/openhome/test_update.py

```diff
@@ -0,0 +1,106 @@
+"""Tests for the Openhome update platform."""
+from unittest.mock import AsyncMock
+
+from homeassistant.components.openhome.update import OpenhomeUpdateEntity
+
+LATEST_FIRMWARE_INSTALLED = {
+    "status": "on_latest",
+    "current_software": {"version": "4.100.502", "topic": "main", "channel": "release"},
+}
+
+FIRMWARE_UPDATE_AVAILABLE = {
+    "status": "update_available",
+    "current_software": {"version": "4.99.491", "topic": "main", "channel": "release"},
+    "update_info": {
+        "legal": {
+            "licenseurl": "http://products.linn.co.uk/VersionInfo/licenseV2.txt",
+            "privacyurl": "https://www.linn.co.uk/privacy",
+            "privacyuri": "https://products.linn.co.uk/VersionInfo/PrivacyV1.json",
+            "privacyversion": 1,
+        },
+        "releasenotesuri": "http://docs.linn.co.uk/wiki/index.php/ReleaseNotes",
+        "updates": [
+            {
+                "channel": "release",
+                "date": "07 Jun 2023 12:29:48",
+                "description": "Release build version 4.100.502 (07 Jun 2023 12:29:48)",
+                "exaktlink": "3",
+                "manifest": "https://cloud.linn.co.uk/update/components/836/4.100.502/manifest.json",
+                "topic": "main",
+                "variant": "836",
+                "version": "4.100.502",
+            }
+        ],
+        "exaktUpdates": [],
+    },
+}
+
+
+async def test_on_latest_firmware():
+    """Test device on latest firmware."""
+    mock_device = AsyncMock()
+
+    mock_device.software_status.return_value = LATEST_FIRMWARE_INSTALLED
+    entity = OpenhomeUpdateEntity(mock_device)
+    await entity.async_update()
+
+    assert entity.installed_version == "4.100.502"
+    assert entity.latest_version is None
+    assert entity.release_url is None
+    assert entity.release_summary is None
+
+
+async def test_update_available():
+    """Test device has firmware update available."""
+    mock_device = AsyncMock()
+
+    mock_device.software_status.return_value = FIRMWARE_UPDATE_AVAILABLE
+    entity = OpenhomeUpdateEntity(mock_device)
+    await entity.async_update()
```

Comment:
Don't test an entity by instantiating it. Instead, forward the config entry to the platform to set it up.

Reply:
I'm really struggling to get this to work using config entry to set it up. I've added the following to get an entry:

```python
async def setup_integration(
    hass: HomeAssistant
) -> MockConfigEntry:
    """Load n openhome platform with mocked device."""
    entry = MockConfigEntry(
        domain=DOMAIN,
        data={CONF_HOST: "http://localhost"},
    )
    entry.add_to_hass(hass)

    with patch("openhomedevice.device.Device", AsyncMock()):
        await hass.config_entries.async_setup(entry.entry_id)
        await hass.async_block_till_done()

    return entry
```

However, the mock just isn't working correctly as it still tries to make network calls unless I stub the `init()` function. Lastly I'm unsure how to get the entity id so I can fetch the state from `hass.states`

Do you know of a good example which mocks an API, calls async update and assert values?

Reply:
You are importing the `Device` class into the package of your integration. You should patch `homeassistant.components.openhome.Device` instead.

Reply:
Thanks - I have it working now. During the construction of the `device_info` object was accessing some private attributes in the `Device` class which was failing during mocking. I've bumped the `openhomedevice` library to 2.2 and made these attributes available through functions. 

The tests now assert the `hass.states` and call the update install service to ensure the integration is wired up correctly.